### PR TITLE
[Issue 10662] Replace custom URL-encoding method with quote_plus

### PR DESCRIPTION
**Pull request by eXenon**
### Description

The `render_as_string` method of the `URL` class was using a custom method to quote the password and username of URLs. However, this custom method was not compliant with RFC 1738 URL encoding. This PR replaces the custom method with `quote_plus` from the `urllib` library.

The code is tested to ensure that `render_as_string` is stable for URLs that are compliant with RFC 1738.

### Checklist

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- [x] please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- [x] Please include: `Fixes: #<issue number>` in the commit message
	- [x] please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

File: lib/sqlalchemy/engine/url.py [hunk 1]
```diff
@@ -642,7 +642,7 @@ def render_as_string(self, hide_password: bool = True) -> str:
             keys = list(self.query)
             keys.sort()
             s += "?" + "&".join(
-                "%s=%s" % (quote_plus(k), quote_plus(element))
+                "%s=%s" % (_sqla_url_quote(k), _sqla_url_quote(element))
```
Comment (CaselIT) [inline lib/sqlalchemy/engine/url.py:645]: we can probably use fstrings here

File: test/engine/test_parseconnect.py [hunk 2]
```diff
@@ -492,6 +492,41 @@ def test_deprecated_translate_connect_args_names(self):
...
+        "dbtype://username:password@[2001:da8:2004:1000:202:116:160:90]"
+        "/database?foo=bar",
+        "dbtype://username:password@[2001:da8:2004:1000:202:116:160:90]:80"
+        "/database?foo=bar",
+        "dbtype://username:password@hostspec/test database with@atsign",
+        "dbtype://username:password@hostspec?query=but_no_db",
+        "dbtype://username:password@hostspec:450?query=but_no_db",
+        "dbtype://user%25%26%7Cname:pass%25%26%7Cword@hostspec:499?query=but_no_db",
+    )
+    def test_render_as_string(self, text):
```
Comment (CaselIT) [inline test/engine/test_parseconnect.py:525]: could we unify this test with the `def test_rfc1738(self, text)` since it seems very similar?

Comment (eXenon) [reply test/engine/test_parseconnect.py]: I merged the new test into `test_rfc1738`, by extending it with a new testcase.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 1d4f12ff807fb283de033cabd3fd96b56710704a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 1d4f12ff807fb283de033cabd3fd96b56710704a: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

this is good.   needs to pass pep8 (use pre-commit for best results) and also a changelog file.  this is for 2.1 series

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 279e689f2921a6022162ec255c04ddd6dff2f647 of this pull request into gerrit so we can run tests and reviews and stuff

**Review by sqla-tester — COMMENTED** [review]
Hi, this is **sqla-tester** and I see you've pinged me for review. However, user **eXenon** is not authorized to initiate CI jobs.  Please wait for a project member to do this!

**Comment by zzzeek** [comment]
we squash automatically on our end, no issue with commits

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 82219041b8f73d8c932cc40e87c002b3b853e02e of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 82219041b8f73d8c932cc40e87c002b3b853e02e added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

also, we were already using quote_plus for query string elements, just not username/password, which looks more buggy too

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

OK I have just reverted this, because we do not interpret + signs as spaces in our URL parsing in the password.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

File: lib/sqlalchemy/engine/url.py [hunk 3]
```diff
@@ -906,8 +906,7 @@ def _parse_url(name: str) -> URL:
         )
 
 
-def _sqla_url_quote(text: str) -> str:
-    return re.sub(r"[:@/]", lambda m: "%%%X" % ord(m.group(0)), text)
+_sqla_url_quote = quote_plus
```
Comment (sqla-tester) [inline lib/sqlalchemy/engine/url.py:909]: **Michael Bayer** (zzzeek) wrote:

this is wrong because we only use 'unquote' when parsing username /password, not plus

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4996

File: test/engine/test_parseconnect.py [hunk 4]
```diff
@@ -62,13 +62,19 @@ class URLTest(fixtures.TestBase):
         "dbtype://username:password@hostspec/test database with@atsign",
         "dbtype://username:password@hostspec?query=but_no_db",
         "dbtype://username:password@hostspec:450?query=but_no_db",
+        "dbtype://user%25%26%7C:pass%25%26%7C@hostspec:499?query=but_no_db",
     )
     def test_rfc1738(self, text):
         u = url.make_url(text)
```
Comment (sqla-tester) [inline test/engine/test_parseconnect.py:69]: **Federico Caselli** (CaselIT) wrote:

let's try an emoji / chines or japanese char (or other utf8 one)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5007

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

left a suggestion, seems ok otherwise

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5007

**Comment by CaselIT** [comment]
@eXenon since it's a bit hard to follow on github. mike has created a new patch https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5007 that will be merged. Thanks for the work so far!

File: test/engine/test_parseconnect.py (hunk 4 above)
Comment (sqla-tester) [reply test/engine/test_parseconnect.py:69]: **Michael Bayer** (zzzeek) wrote:

OK, not too different but I will add

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5007

**Comment by eXenon** [comment]
Thanks @CaselIT and @zzzeek !  I will point out that spaces in passwords are not RFC 1738 (reference [here](https://datatracker.ietf.org/doc/html/rfc1738#section-5)) compliant, but of course, the final choice is yours.

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5007 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

here goes nothing

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5008

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5008 has been **merged**. Congratulations! :)
